tests/koji/test_koji_vmi.py

```diff
@@ -0,0 +1,204 @@
+import os
+
+from pytest import raises
+
+from pushsource import (
+    Source,
+    AmiPushItem,
+    VHDPushItem,
+    VMIPushItem,
+    AmiRelease,
+    VMIRelease,
+)
+
+
+def test_koji_vmi_notfound(fake_koji):
+    """koji source referencing nonexistent VMI build will fail."""
+
+    source = Source.get("koji:https://koji.example.com/?vmi_build=notexist-1.2.3")
+    with raises(ValueError) as exc_info:
+        list(source)
+
+    assert "Virtual machine image build not found in koji: notexist-1.2.3" in str(
+        exc_info.value
+    )
+
+
+def test_koji_not_vmi(fake_koji, koji_dir):
+    """Koji source referencing an invalid build will fail."""
+
+    source = Source.get(
+        "koji:https://koji.example.com/?vmi_build=foobuild-1.0-1", basedir=koji_dir
+    )
+
+    fake_koji.build_data[1234] = {
+        "id": 1234,
+        "name": "foobuild",
+        "version": "1.0",
+        "release": "1",
+        "nvr": "foobuild-1.0-1",
+        "completion_time": "2022-12-20 12:30:25",
+    }
+    fake_koji.build_data["foobuild-1.0-1"] = fake_koji.build_data[1234]
+
+    with raises(ValueError) as exc_info:
+        list(source)
+
+    assert (
+        "Build foobuild-1.0-1 not recognized as a virtual machine image build"
+        in str(exc_info.value)
+    )
+
+
+def test_koji_vmis(fake_koji, koji_dir):
+    """Koji source yields requested Virtual Machine Images."""
+
+    nvr = "foobuild-1.0-1"
+    r, v, n = nvr[-1::-1].split("-", 2)
```

Comment:
I think it would be better to just write out the assignments of `name`, `version`, `release` in full...

There is after all no need for this code to support arbitrary NVRs since the only handled value is hardcoded anyway.  It would be easier to read if it simply said name="foobuild", version="1.0", release="1" and would reduce the possibility of the test itself being buggy.

Reply:
Ok, I'll change it

Reply:
Fixed! It makes more sense to hardcode as you suggested, thanks a lot!